crates/executor/src/executor.rs

```diff
@@ -17,7 +18,157 @@ use revm::{
+where
+    DB: StateProvider,
+{
+    chain_spec: &'a ChainSpec,
+    evm: EVM<&'a mut SubState<DB>>,
+}
+
+impl<'a, DB> Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn new(chain_spec: &'a ChainSpec, db: &'a mut SubState<DB>) -> Self {
+        let mut evm = EVM::new();
+        evm.database(db);
+        Executor { chain_spec, evm }
+    }
+
+    fn recover_signers(
+        &self,
+        body: &[TransactionSigned],
+        signers: Option<Vec<Address>>,
+    ) -> Result<Vec<Address>, Error> {
+        if let Some(signers) = signers {
+            if body.len() == signers.len() {
+                Ok(signers)
+            } else {
+                Err(Error::SignerRecoveryError)
+            }
+        } else {
+            body.iter().map(|tx| tx.recover_signer().ok_or(Error::SignerRecoveryError)).collect()
+        }
+    }
+
+    fn init_block_env(&mut self, header: &Header) {
+        let spec_id = revm_spec(self.chain_spec, header.number);
+        self.evm.env.cfg.chain_id = U256::from(self.chain_spec.chain().id());
+        self.evm.env.cfg.spec_id = spec_id;
+        self.evm.env.cfg.perf_all_precompiles_have_balance = false;
+        self.evm.env.cfg.perf_analyse_created_bytecodes = AnalysisKind::Raw;
+
+        revm_wrap::fill_block_env(&mut self.evm.env.block, header, spec_id >= SpecId::MERGE);
+    }
+}
+
+impl<'a, DB> BlockExecutor for Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn execute(
+        &mut self,
+        block: &Block,
+        signers: Option<Vec<Address>>,
+    ) -> Result<ExecutionResult, Error> {
+        let Block { header, body, ommers } = block;
+        let signers = self.recover_signers(body, signers)?;
+
+        self.init_block_env(header);
+
+        let mut cumulative_gas_used = 0;
+        // output of verification
+        let mut changesets = Vec::with_capacity(body.len());
+
+        for (transaction, signer) in body.iter().zip(signers.into_iter()) {
+            // The sum of the transaction’s gas limit, Tg, and the gas utilised in this block prior,
+            // must be no greater than the block’s gasLimit.
+            let block_available_gas = header.gas_limit - cumulative_gas_used;
+            if transaction.gas_limit() > block_available_gas {
+                return Err(Error::TransactionGasLimitMoreThenAvailableBlockGas {
+                    transaction_gas_limit: transaction.gas_limit(),
+                    block_available_gas,
+                })
+            }
+
+            // Fill revm structure.
+            revm_wrap::fill_tx_env(&mut self.evm.env.tx, transaction, signer);
+
+            // Execute transaction.
+            let out = self.evm.transact();
+
+            // Useful for debugging
+            // let out = evm.inspect(revm::inspectors::CustomPrintTracer::default());
+            // tracing::trace!(target:"evm","Executing transaction {:?}, \n:{out:?}: {:?}
+            // \nENV:{:?}",transaction.hash(),transaction,evm.env);
```

Comment:
We should either remove this or add it as an optional feature of the crate: https://doc.rust-lang.org/cargo/reference/features.html

Reply:
Before going to bed, I'd like to quick response it. This comment block is originally existed on `execute()`. I didn't fully understand why it's useful so I just kept. If I should decide it, I'd like to remove it but I need more feedback.

Reply:
it was used for debugging and can be useful for me. It is going to be removed on its own when we implement support for traces

Reply:
@onbjerg We need to keep this commented line. Could we mark it as resolved?

Reply:
Let's keep it around given Dragan uses for debugging, and we can remove later.